### PR TITLE
Python: `detect_constraint_structure`

### PR DESCRIPTION
**Pull request by aghezz1**

### CODE REVIEW

File: interfaces/acados_template/acados_template/mpc_utils.py [hunk 1]
```diff
@@ -117,3 +118,194 @@ def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarra
     model.f_impl_expr = ca.vertcat(model.f_impl_expr, cost_state_dot-cost_dot)
 
     return model, ocp.parameter_values
+
+
+def detect_constr(model: AcadosModel, constraints: AcadosOcpConstraints, stage_type: str):
```
[the remark this answers is not in the document]
Comment (aghezz1) [reply interfaces/acados_template/acados_template/mpc_utils.py:123]: Agree, I named it to keep similarity with matlab interface.
We might consider to name both `detect_constraints`

File: interfaces/acados_template/acados_template/mpc_utils.py [hunk 2]
```diff
@@ -117,3 +118,198 @@ def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarra
...
+    u = model.u
+    z = model.z
+    nx = x.shape[0]
+    nu = u.shape[0]
+    nz = z.shape[0]
+
+    if isinstance(x, ca.SX):
+        isSX = True
+    else:
+        raise ValueError('Constraint detection only works for casadi.SX!')
```
Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/mpc_utils.py:137]: what is missing to have this work with MX?

File: interfaces/acados_template/acados_template/mpc_utils.py [hunk 3]
```diff
@@ -117,3 +118,198 @@ def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarra
...
+    ubx = []
+
+    Jbu = ca.SX.zeros(0, nu)
+    lbu = []
+    ubu = []
+
+    # Loop over CasADi formulated constraints
+    for ii in range(expr_constr.shape[0]):
+        c = expr_constr[ii]
+        # TODO or any(ca.which_depends(c, model.p_global)):
```
Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/mpc_utils.py:186]: this TODO can be removed I guess?

**Comment by sandmaennchen** [comment]
and could you add the check for `z` and `p_global` in MATLAB, too? :blush:

File: interfaces/acados_template/acados_template/mpc_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (FreyJo) [reply interfaces/acados_template/acados_template/mpc_utils.py:123]: You can rename both.
In Matlab the function is only called in the interface, not by the user, so renaming is not a breaking change.

[the remark this answers is not in the document]
Comment (aghezz1) [reply interfaces/acados_template/acados_template/mpc_utils.py:123]: Ok, I will rename them and maybe set up some tests.

File: interfaces/acados_template/acados_template/mpc_utils.py (hunk 2 above)
Comment (aghezz1) [reply interfaces/acados_template/acados_template/mpc_utils.py:137]: Not sure, I was copying/pasting from the matlab files.
I will test if also MX are supported.

File: interfaces/acados_template/acados_template/mpc_utils.py [hunk 4]
```diff
@@ -79,41 +81,250 @@ def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarra
...
+    x = model.x
+    u = model.u
+    z = model.z
+    nx = x.shape[0]
+    nu = u.shape[0]
+
+    if model.p_global is None:
+        p_global = []  # to have same structure of model.p
+
+    if isinstance(x, ca.SX):
```
Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/mpc_utils.py:147]: you can use `model.get_casadi_symbol()` here

Comment (FreyJo) [reply interfaces/acados_template/acados_template/mpc_utils.py:147]: I think he also needs the zeros though.

File: interfaces/acados_template/acados_template/mpc_utils.py [hunk 5]
```diff
@@ -79,41 +81,250 @@ def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarra
...
+        CASADI_VAR = ca.SX
+    elif isinstance(x, ca.MX):
+        CASADI_VAR = ca.MX
+    else:
+        raise TypeError(
+            "Optimization variables 'x' must be of type 'casadi SX' or 'casadi MX'")
+
+    if stage_type == 'initial':
+        expr_constr = model.con_h_expr_0
+        LB = constraints.lh_0
```
Comment (FreyJo) [inline interfaces/acados_template/acados_template/mpc_utils.py:157]: I would avoid all upper case, in Python this should be used for global variables.

File: interfaces/acados_template/acados_template/mpc_utils.py [hunk 6]
```diff
@@ -144,13 +144,8 @@ def detect_constraint_structure(model: AcadosModel, constraints: AcadosOcpConstr
...
 
-    if isinstance(x, ca.SX):
-        casadi_var = ca.SX
-    elif isinstance(x, ca.MX):
-        casadi_var = ca.MX
-    else:
-        raise TypeError(
-            "Optimization variables 'x' must be of type 'casadi SX' or 'casadi MX'")
+    casadi_var = model.get_casadi_symbol()
+    casadi_dm_zeros = ca.DM.zeros
```
Comment (aghezz1) [inline interfaces/acados_template/acados_template/mpc_utils.py:148]: I used `get_casadi_symbol()` for creating symbols but for the zeros I needed casadi DM so I did it as above

**Review by FreyJo — APPROVED** [review]

**Pull request merged**